Code/GraphMol/AddHs.cpp

```diff
@@ -1263,5 +1269,50 @@ bool needsHs(const ROMol &mol) {
   return false;
 }
 
+bool hasQueryHs(const ROMol &mol, bool unmergableOnly) {
+  int unmergableHs = 0;
+  int mergableHs = 0;
+  // We don't care about announcing ORs or other items during isQueryH
+  RDLog::LogStateSetter blocker;
+
+  for (const auto &atom : mol.atoms()) {
+    switch (isQueryH(atom)) {
+      case HydrogenType::QueryHydrogen:
+        mergableHs += 1;
+        break;
+      case HydrogenType::UnMergableQueryHydrogen:
+        unmergableHs += 1;
+        break;
+      case HydrogenType::NotAHydrogen:
+        break;
+    }
+    if (atom->hasQuery()) {
+      if (atom->getQuery()->getDescription() == "RecursiveStructure") {
+        auto *rsq = dynamic_cast<RecursiveStructureQuery *>(atom->getQuery());
+        CHECK_INVARIANT(rsq, "could not convert recursive structure query");
+        if (hasQueryHs(*rsq->getQueryMol(), unmergableOnly)) return true;
+      }
+
+      // FIX: shouldn't be repeating this code here -- yet again!
+      std::list<QueryAtom::QUERYATOM_QUERY::CHILD_TYPE> childStack(
+          atom->getQuery()->beginChildren(), atom->getQuery()->endChildren());
+      while (childStack.size()) {
+        QueryAtom::QUERYATOM_QUERY::CHILD_TYPE qry = childStack.front();
+        childStack.pop_front();
+        if (qry->getDescription() == "RecursiveStructure") {
+          auto *rsq = dynamic_cast<RecursiveStructureQuery *>(qry.get());
+          CHECK_INVARIANT(rsq, "could not convert recursive structure query");
+          if (hasQueryHs(*rsq->getQueryMol(), unmergableOnly)) return true;
+        } else if (qry->beginChildren() != qry->endChildren()) {
```

Comment:
It doesn't hurt to have it, but I don't think the `if` is actually needed here
```suggestion
        } else {
```